main/src/ca/uwaterloo/flix/api/lsp/provider/HighlightProvider.scala

```diff
@@ -15,195 +16,369 @@
+
+      case _ => mkNotFound(uri, pos)
+    }
+  }
 
-        case Entity.Def(defn) => highlightDef(uri, defn.sym)
+  private def highlightLabel(label: Name.Label)(implicit root: Root, acceptor: Visitor.Acceptor): JObject = {
+    val builder = HighlightBuilder(label)
 
-        case Entity.Sig(sig0) => highlightSig(uri, sig0.sym)
+    object LabelConsumer extends Consumer {
+      override def consumeExpr(exp: Expr): Unit = exp match {
+        case Expr.RecordExtend(l, _, _, _, _, _) => builder.considerWrite(l, l.loc)
+        case Expr.RecordSelect(_, l, _, _, _) => builder.considerRead(l, l.loc)
+        case Expr.RecordRestrict(l, _, _, _, _) => builder.considerWrite(l, l.loc)
+        case _ => ()
+      }
+    }
 
-        case Entity.Enum(enum0) => highlightEnum(uri, enum0.sym)
+    Visitor.visitRoot(root, LabelConsumer, acceptor)
 
-        case Entity.Struct(struct) => highlightStruct(uri, struct.sym)
+    builder.build
+  }
 
-        case Entity.TypeAlias(alias) => highlightTypeAlias(uri, alias.sym)
+  private def highlightTypeVarSym(sym: Symbol.KindedTypeVarSym)(implicit root: Root, acceptor: Visitor.Acceptor): JObject = {
+    val builder = HighlightBuilder(sym)
 
-        case Entity.AssocType(assoc) => highlightAssocType(uri, assoc.sym)
+    object TypeVarSymConsumer extends Consumer {
+      override def consumeTypeParam(tparam: TypedAst.TypeParam): Unit = builder.considerWrite(tparam.sym, tparam.sym.loc)
+      override def consumeType(tpe: Type): Unit = tpe match {
+        case Type.Var(sym, loc) => builder.considerRead(sym, loc)
+        case _ => ()
+      }
+    }
 
-        case Entity.Effect(eff) => highlightEffect(uri, eff.sym)
+    Visitor.visitRoot(root, TypeVarSymConsumer, acceptor)
 
-        case Entity.Op(op) => highlightOp(uri, op.sym)
+    builder.build
+  }
 
-        case Entity.VarUse(sym, _, _) => highlightVar(uri, sym)
+  private def highlightAssocTypeSym(sym: Symbol.AssocTypeSym)(implicit root: Root, acceptor: Visitor.Acceptor): JObject = {
+    val builder = HighlightBuilder(sym)
 
-        case Entity.DefUse(sym, _, _) => highlightDef(uri, sym)
+    object AssocTypeSymConsumer extends Consumer {
+      override def consumeAssocTypeSig(tsig: TypedAst.AssocTypeSig): Unit = builder.considerWrite(tsig.sym, tsig.sym.loc)
+      override def consumeAssocTypeSymUse(symUse: SymUse.AssocTypeSymUse): Unit = builder.considerRead(symUse.sym, symUse.loc)
+    }
 
-        case Entity.SigUse(sym, _, _) => highlightSig(uri, sym)
+    Visitor.visitRoot(root, AssocTypeSymConsumer, acceptor)
 
-        case Entity.CaseUse(sym, _, _) => highlightCase(uri, sym)
+    builder.build
+  }
+
+  /**
+    * A builder for creating an LSP highlight response containing a [[DocumentHighlight]] for each recorded
+    * occurrence of a given token [[tok]].
+    *
+    * An occurrence of a token of type [[T]] is "considered" by invoking [[considerRead]] or [[considerWrite]] for resp.
```

Comment:
The word token has a very specific meaning in Flix. What is this token?

Reply:
It used to be Symbol, when we were just finding occurrences of symbols. But now were findind occurrences of labels as well. So it's more like an sequence of letters that mark or name some unit that might occur repeatedly. Token was the best name I could come up with, but I'm open to alternatives. Maybe identifier or something?